File: src/main/java/com/vc/hard/L668.scala
package com.vc.hard

object L668 {
  def findKthNumber(m: Int, n: Int, k: Int): Int = {
    var lo = 0
    var hi = m * n

    while(lo <= hi) {
      val mid = lo + (hi - lo) / 2

      var row = 1
      var col = n
      var count = 0
      while(row <= m && col > 0) {
        //Take full row
        if(row * col <= mid) {
          count += col
          row += 1
        }
        //decrement col until you find number less than or equal to mid
        else col -= 1
      }

      if(count < k) lo = mid + 1
      else hi = mid - 1
    }
    lo
  }
}

File: src/main/java/com/vc/hard/L668.java
package com.vc.hard;

class L668 {
    public int findKthNumber(int m, int n, int k) {
        int lo = 0;
        int hi = m * n;
        while(lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            int count = count(m, n, mid);
            if(count < k) lo = mid + 1;
            else hi = mid - 1;
        }
        return lo;
    }

    private int count(int m, int n, int number) {
        int row = 1;
        int col = n;
        int count = 0;
        while(row <= m && col > 0) {
            if(row * col <= number) {
                count += col;
                row++;
            }
            else col--;
        }
        return count;
    }
}
